test/metabase/permissions_test.clj

```diff
@@ -0,0 +1,743 @@
+                    Segment [db2-expensive-venues (expensive-venues-segment "DB 2 Expensive Venues" (table *db2* :venues))]
+                    Segment [db2-todays-users     (todays-users-segment     "DB 2 Today's Users"    (table *db2* :users))]]
+      (binding [*segment:db1-expensive-venues* db1-expensive-venues
+                *segment:db2-expensive-venues* db2-expensive-venues
+                *segment:db2-todays-users*     db2-todays-users]
+        (f)))))
+
+;;; ------------------------------------------------------------ with everything! ------------------------------------------------------------
+
+
+(defn- -do-with-test-data [f]
+  (((comp with-ops-group
+          with-db-2
+          with-db-1
+          with-cards
+          with-dashboards
+          with-pulses
+          with-metrics
+          with-segments) f)))
+
+(defmacro ^:private with-test-data {:style/indent 0} [& body]
+  `(-do-with-test-data (fn []
+                         ~@body)))
+
+(defmacro ^:private expect-with-test-data {:style/indent 0} [expected actual]
+  `(expect
+     ~expected
+     (with-test-data
+       ~actual)))
+
+
+;;; +----------------------------------------------------------------------------------------------------------------------------------------------------------------+
+;;; |                                                                          QUERY BUILDER                                                                         |
+;;; +----------------------------------------------------------------------------------------------------------------------------------------------------------------+
+
+;;; ------------------------------------------------------------ GET /api/database?include_tables=true (Visible DBs + Tables) ------------------------------------------------------------
+
+(defn- GET-database [username]
+  (vec (for [db    ((test-users/user->client username) :get 200 "database", :include_tables true)
+             :when (contains? (all-db-ids) (u/get-id db))]
+         [(:name db) (mapv :name (:tables db))])))
+
+;; admin should be able to see everything
+(expect-with-test-data
+  [["DB One" ["checkins" "users" "venues"]]
+   ["DB Two" ["checkins" "users" "venues"]]]
+  (GET-database :crowberto))
+
+;; basic user should only see DB 1
+(expect-with-test-data
+  [["DB One" ["checkins" "users" "venues"]]]
+  (GET-database :rasta))
+
+;; ops user should see DB 1 and venues in DB 2
+(expect-with-test-data
+  [["DB One" ["checkins" "users" "venues"]]
+   ["DB Two" ["venues"]]]
+  (GET-database :lucky))
+
+
+;;; ------------------------------------------------------------ GET /api/table/:id/query_metadata ------------------------------------------------------------
+
+(defn- GET-table-query-metadata [username db table-name]
+  (not= ((test-users/user->client username) :get (format "table/%d/query_metadata" (u/get-id (table db table-name))))
+        "You don't have permissions to do that."))
```

Comment:
Yeah it is throwing a 403, the "You don't have permissions to do that." message is the specific response we always return for 403s

Unfortunately the way the tests are set up it's not easy to test for one status code versus a different one, that would require reworking the HTTP test client code